From 6419f78ff22d42e69a6877bb6e691a533e340595 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Thu, 11 Jun 2026 08:31:09 +0200
Subject: [PATCH] [core,gateway] rpc header max fragment size checks

---
 libfreerdp/core/gateway/rpc_bind.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/libfreerdp/core/gateway/rpc_bind.c b/libfreerdp/core/gateway/rpc_bind.c
index bdf54d5a0..e4d78b0dd 100644
--- a/libfreerdp/core/gateway/rpc_bind.c
+++ b/libfreerdp/core/gateway/rpc_bind.c
@@ -335,6 +335,17 @@ BOOL rpc_recv_bind_ack_pdu(rdpRpc* rpc, wStream* s)
 
 	WLog_DBG(TAG, "Receiving BindAck PDU");
 
+	const UINT16 MAX_VALID_FRAG = 0x0FF8;
+	if ((header.bind_ack.max_xmit_frag > MAX_VALID_FRAG) ||
+	    (header.bind_ack.max_recv_frag > MAX_VALID_FRAG))
+	{
+		WLog_ERR(TAG,
+		         "bind_ack: invalid fragment size: max_xmit_frag=%" PRIu16
+		         ", max_recv_frag=%" PRIu16 ", maximum=%" PRIu16,
+		         header.bind_ack.max_xmit_frag, header.bind_ack.max_recv_frag, MAX_VALID_FRAG);
+		goto fail;
+	}
+
 	rpc->max_recv_frag = header.bind_ack.max_xmit_frag;
 	rpc->max_xmit_frag = header.bind_ack.max_recv_frag;
 
