From 5f407b0a8c1d6e913476d6063e8d4c70bc1406b5 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Wed, 19 Aug 2026 09:13:23 +0200
Subject: [PATCH] [core,nego] fix capacity checks

---
 libfreerdp/core/nego.c | 32 ++++++++++++++++++++------------
 1 file changed, 20 insertions(+), 12 deletions(-)

diff --git a/libfreerdp/core/nego.c b/libfreerdp/core/nego.c
index 5c3c54a11..b5db255aa 100644
--- a/libfreerdp/core/nego.c
+++ b/libfreerdp/core/nego.c
@@ -851,7 +851,6 @@ BOOL nego_send_negotiation_request(rdpNego* nego)
 {
 	BOOL rc = FALSE;
 	wStream* s;
-	size_t length;
 	size_t bm, em;
 	BYTE flags = 0;
 	size_t cookie_length;
@@ -863,12 +862,14 @@ BOOL nego_send_negotiation_request(rdpNego* nego)
 		return FALSE;
 	}
 
-	length = TPDU_CONNECTION_REQUEST_LENGTH;
 	bm = Stream_GetPosition(s);
-	Stream_Seek(s, length);
+	if (!Stream_SafeSeek(s, TPDU_CONNECTION_REQUEST_LENGTH))
+		return FALSE;
 
 	if (nego->RoutingToken)
 	{
+		if (!Stream_EnsureRemainingCapacity(s, nego->RoutingTokenLength))
+			return FALSE;
 		Stream_Write(s, nego->RoutingToken, nego->RoutingTokenLength);
 
 		/* Ensure Routing Token is correctly terminated - may already be present in string */
@@ -878,14 +879,14 @@ BOOL nego_send_negotiation_request(rdpNego* nego)
 		    (nego->RoutingToken[nego->RoutingTokenLength - 1] == 0x0A))
 		{
 			WLog_DBG(TAG, "Routing token looks correctly terminated - use verbatim");
-			length += nego->RoutingTokenLength;
 		}
 		else
 		{
 			WLog_DBG(TAG, "Adding terminating CRLF to routing token");
+			if (!Stream_EnsureRemainingCapacity(s, 2))
+				return FALSE;
 			Stream_Write_UINT8(s, 0x0D); /* CR */
 			Stream_Write_UINT8(s, 0x0A); /* LF */
-			length += nego->RoutingTokenLength + 2;
 		}
 	}
 	else if (nego->cookie)
@@ -895,11 +896,16 @@ BOOL nego_send_negotiation_request(rdpNego* nego)
 		if (cookie_length > nego->CookieMaxLength)
 			cookie_length = nego->CookieMaxLength;
 
+		if (!Stream_EnsureRemainingCapacity(s, 17))
+			return FALSE;
 		Stream_Write(s, "Cookie: mstshash=", 17);
+		if (!Stream_EnsureRemainingCapacity(s, cookie_length))
+			return FALSE;
 		Stream_Write(s, (BYTE*)nego->cookie, cookie_length);
+		if (!Stream_EnsureRemainingCapacity(s, 2))
+			return FALSE;
 		Stream_Write_UINT8(s, 0x0D); /* CR */
 		Stream_Write_UINT8(s, 0x0A); /* LF */
-		length += cookie_length + 19;
 	}
 
 	WLog_DBG(TAG, "RequestedProtocols: %" PRIu32 "", nego->RequestedProtocols);
@@ -910,20 +916,22 @@ BOOL nego_send_negotiation_request(rdpNego* nego)
 		if (nego->RestrictedAdminModeRequired)
 			flags |= RESTRICTED_ADMIN_MODE_REQUIRED;
 
+		if (!Stream_EnsureRemainingCapacity(s, 8))
+			return FALSE;
+
 		Stream_Write_UINT8(s, TYPE_RDP_NEG_REQ);
 		Stream_Write_UINT8(s, flags);
 		Stream_Write_UINT16(s, 8);                        /* RDP_NEG_DATA length (8) */
 		Stream_Write_UINT32(s, nego->RequestedProtocols); /* requestedProtocols */
-		length += 8;
 	}
 
-	if (length > UINT16_MAX)
-		goto fail;
-
 	em = Stream_GetPosition(s);
+	if ((em < 5) || (em > UINT16_MAX))
+		goto fail;
 	Stream_SetPosition(s, bm);
-	tpkt_write_header(s, (UINT16)length);
-	tpdu_write_connection_request(s, (UINT16)length - 5);
+	if (!tpkt_write_header(s, (UINT16)em))
+		goto fail;
+	tpdu_write_connection_request(s, (UINT16)em - 5);
 	Stream_SetPosition(s, em);
 	Stream_SealLength(s);
 	rc = (transport_write(nego->transport, s) >= 0);
