From 9dd6b629d3151746bda447db8286429d7443e2f0 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Thu, 11 Jun 2026 08:30:18 +0200
Subject: [PATCH] [core,gateway] ensure capacity in rpc_channel_read

---
 libfreerdp/core/gateway/rpc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libfreerdp/core/gateway/rpc.c b/libfreerdp/core/gateway/rpc.c
index d90d80ad7..dcfed0207 100644
--- a/libfreerdp/core/gateway/rpc.c
+++ b/libfreerdp/core/gateway/rpc.c
@@ -465,6 +465,9 @@ SSIZE_T rpc_channel_read(RpcChannel* channel, wStream* s, size_t length)
 	if (!channel || (length > INT32_MAX))
 		return -1;
 
+	if (!Stream_EnsureRemainingCapacity(s, length))
+		return -1;
+
 	ERR_clear_error();
 	status = BIO_read(channel->tls->bio, Stream_Pointer(s), (INT32)length);
 
