From f2b44427946bf42128609b66b4f6623c9b58a482 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Thu, 30 Apr 2026 10:04:24 +0200
Subject: [PATCH] [gdi,gfx] fix bounds checks

---
 libfreerdp/gdi/gfx.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/libfreerdp/gdi/gfx.c b/libfreerdp/gdi/gfx.c
index 3cbea681d..d22405445 100644
--- a/libfreerdp/gdi/gfx.c
+++ b/libfreerdp/gdi/gfx.c
@@ -1519,10 +1519,12 @@ static UINT gdi_CacheToSurface(RdpgfxClientContext* context,
 		if (!is_rect_valid(&rect, surface->width, surface->height))
 			goto fail;

+		const UINT32 w = rect.right - rect.left;
+		const UINT32 h = rect.bottom - rect.top;
 		if (!freerdp_image_copy_no_overlap(surface->data, surface->format, surface->scanline,
-		                                   destPt->x, destPt->y, cacheEntry->width,
-		                                   cacheEntry->height, cacheEntry->data, cacheEntry->format,
-		                                   cacheEntry->scanline, 0, 0, NULL, FREERDP_FLIP_NONE))
+		                                   destPt->x, destPt->y, w, h, cacheEntry->data,
+		                                   cacheEntry->format, cacheEntry->scanline, 0, 0, NULL,
+		                                   FREERDP_FLIP_NONE))
 			goto fail;

 		invalidRect = rect;
--
2.52.0
