From 9958665d48540747f0a906e3754ef84c3371cbec Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Wed, 15 Jul 2026 08:42:12 +0200
Subject: [PATCH] [winpr,sspi] tighten bounds checks for kerberos

---
 winpr/libwinpr/sspi/Kerberos/kerberos.c | 53 ++++++++++++-------------
 1 file changed, 25 insertions(+), 28 deletions(-)

diff --git a/winpr/libwinpr/sspi/Kerberos/kerberos.c b/winpr/libwinpr/sspi/Kerberos/kerberos.c
index 87335316c..f1a908923 100644
--- a/winpr/libwinpr/sspi/Kerberos/kerberos.c
+++ b/winpr/libwinpr/sspi/Kerberos/kerberos.c
@@ -1894,22 +1894,6 @@ static SECURITY_STATUS SEC_ENTRY kerberos_DecryptMessage(PCtxtHandle phContext,
 {
 #ifdef WITH_KRB5
 	KRB_CONTEXT* context = get_context(phContext);
-	PSecBuffer sig_buffer = NULL;
-	PSecBuffer data_buffer = NULL;
-	krb5glue_key key = NULL;
-	krb5_keyusage usage = 0;
-	char* header = NULL;
-	uint16_t tok_id = 0;
-	BYTE flags = 0;
-	uint16_t ec = 0;
-	uint16_t rrc = 0;
-	uint64_t seq_no = 0;
-	krb5_crypto_iov iov[] = { { KRB5_CRYPTO_TYPE_HEADER, { 0 } },
-		                      { KRB5_CRYPTO_TYPE_DATA, { 0 } },
-		                      { KRB5_CRYPTO_TYPE_DATA, { 0 } },
-		                      { KRB5_CRYPTO_TYPE_PADDING, { 0 } },
-		                      { KRB5_CRYPTO_TYPE_TRAILER, { 0 } } };
-
 	if (!context)
 		return SEC_E_INVALID_HANDLE;
 
@@ -1918,19 +1902,19 @@ static SECURITY_STATUS SEC_ENTRY kerberos_DecryptMessage(PCtxtHandle phContext,
 
 	KRB_CREDENTIALS* creds = context->credentials;
 
-	sig_buffer = sspi_FindSecBuffer(pMessage, SECBUFFER_TOKEN);
-	data_buffer = sspi_FindSecBuffer(pMessage, SECBUFFER_DATA);
+	const PSecBuffer sig_buffer = sspi_FindSecBuffer(pMessage, SECBUFFER_TOKEN);
+	PSecBuffer data_buffer = sspi_FindSecBuffer(pMessage, SECBUFFER_DATA);
 
 	if (!sig_buffer || !data_buffer || sig_buffer->cbBuffer < 16)
 		return SEC_E_INVALID_TOKEN;
 
 	/* Read in header information */
-	header = sig_buffer->pvBuffer;
-	tok_id = winpr_Data_Get_UINT16_BE(header);
-	flags = header[2];
-	ec = winpr_Data_Get_UINT16_BE((header + 4));
-	rrc = winpr_Data_Get_UINT16_BE((header + 6));
-	seq_no = winpr_Data_Get_UINT64_BE((header + 8));
+	char* header = sig_buffer->pvBuffer;
+	const uint16_t tok_id = winpr_Data_Get_UINT16_BE(header);
+	const BYTE flags = header[2];
+	const uint16_t ec = winpr_Data_Get_UINT16_BE((header + 4));
+	const uint16_t rrc = winpr_Data_Get_UINT16_BE((header + 6));
+	const uint64_t seq_no = winpr_Data_Get_UINT64_BE((header + 8));
 
 	/* Check that the header is valid */
 	if (tok_id != TOK_ID_WRAP || (BYTE)header[3] != 0xFF)
@@ -1951,12 +1935,17 @@ static SECURITY_STATUS SEC_ENTRY kerberos_DecryptMessage(PCtxtHandle phContext,
 		return SEC_E_INVALID_TOKEN;
 
 	/* Find the proper key and key usage */
-	key = get_key(&context->keyset);
+	krb5glue_key key = get_key(&context->keyset);
 	if (!key || ((flags & FLAG_ACCEPTOR_SUBKEY) && (context->keyset.acceptor_key != key)))
 		return SEC_E_INTERNAL_ERROR;
-	usage = context->acceptor ? KG_USAGE_INITIATOR_SEAL : KG_USAGE_ACCEPTOR_SEAL;
+	krb5_keyusage usage = context->acceptor ? KG_USAGE_INITIATOR_SEAL : KG_USAGE_ACCEPTOR_SEAL;
 
 	/* Fill in the lengths of the iov array */
+	krb5_crypto_iov iov[] = { { KRB5_CRYPTO_TYPE_HEADER, { 0 } },
+		                      { KRB5_CRYPTO_TYPE_DATA, { 0 } },
+		                      { KRB5_CRYPTO_TYPE_DATA, { 0 } },
+		                      { KRB5_CRYPTO_TYPE_PADDING, { 0 } },
+		                      { KRB5_CRYPTO_TYPE_TRAILER, { 0 } } };
 	iov[1].data.length = data_buffer->cbBuffer;
 	iov[2].data.length = 16;
 	if (krb_log_exec(krb5glue_crypto_length_iov, creds->ctx, key, iov, ARRAYSIZE(iov)))
@@ -1969,9 +1958,17 @@ static SECURITY_STATUS SEC_ENTRY kerberos_DecryptMessage(PCtxtHandle phContext,
 		return SEC_E_INVALID_TOKEN;
 
 	/* Locate the parts of the message */
-	iov[0].data.data = header + 16 + rrc + ec;
+	const size_t iov0Offset = 16ull + rrc + ec;
+	if (iov0Offset + iov[0].data.length > sig_buffer->cbBuffer)
+		return SEC_E_INVALID_TOKEN;
+
+	const size_t iov2Offset = 16ull + ec;
+	if (iov2Offset + iov[2].data.length > sig_buffer->cbBuffer)
+		return SEC_E_INVALID_TOKEN;
+
+	iov[0].data.data = header + iov0Offset;
 	iov[1].data.data = data_buffer->pvBuffer;
-	iov[2].data.data = header + 16 + ec;
+	iov[2].data.data = header + iov2Offset;
 	iov[3].data.data = iov[2].data.data + iov[2].data.length;
 	iov[4].data.data = iov[3].data.data + iov[3].data.length;
 
